core/src/test/scala/unit/kafka/network/ConnectionQuotasTest.scala

```diff
@@ -0,0 +1,307 @@
+    val props = TestUtils.createBrokerConfig(0, TestUtils.MockZkConnect, port = 0)
+    props.put(KafkaConfig.ListenersProp, "EXTERNAL://localhost:0,REPLICATION://localhost:1,ADMIN://localhost:2")
+    // ConnectionQuotas does not limit inter-broker listener even when broker-wide connection limit is reached
+    props.put(KafkaConfig.InterBrokerListenerNameProp, "REPLICATION")
+    props.put(KafkaConfig.ListenerSecurityProtocolMapProp, "EXTERNAL:PLAINTEXT,REPLICATION:PLAINTEXT,ADMIN:PLAINTEXT")
+    props
+  }
+
+  @Before
+  def setUp(): Unit = {
+    // Clean-up any metrics left around by previous tests
+    TestUtils.clearYammerMetrics()
+
+    listeners.keys.foreach { name =>
+        blockedPercentMeters.put(name, KafkaMetricsGroup.newMeter(
+          s"${name}BlockedPercent", "blocked time", TimeUnit.NANOSECONDS, Map(ListenerMetricTag -> name)))
+    }
+  }
+
+  @After
+  def tearDown(): Unit = {
+    TestUtils.clearYammerMetrics()
+    blockedPercentMeters.clear()
+  }
+
+  @Test
+  def testFailWhenNoListeners(): Unit = {
+    val config = KafkaConfig.fromProps(brokerPropsWithDefaultConnectionLimits)
+    val connectionQuotas = new ConnectionQuotas(config, time)
+
+    val executor = Executors.newSingleThreadExecutor
+    try {
+      // inc() on a separate thread in case it blocks
+      val externalListener = listeners("EXTERNAL")
+      executor.submit((() =>
+        intercept[RuntimeException](connectionQuotas.inc(externalListener.listenerName, externalListener.defaultClientIp, blockedPercentMeters("EXTERNAL")))): Runnable
+      ).get(5, TimeUnit.SECONDS)
+    } finally {
+      executor.shutdownNow()
+    }
+  }
+
+  @Test
+  def testNoConnectionLimitsByDefault(): Unit = {
+    val config = KafkaConfig.fromProps(brokerPropsWithDefaultConnectionLimits)
+    val connectionQuotas = new ConnectionQuotas(config, time)
+    addListenersAndVerify(config, connectionQuotas)
+
+    val executor = Executors.newFixedThreadPool(listeners.size)
+    try {
+      // verify there is no limit by accepting 10000 connections as fast as possible
+      val numConnections = 10000
+      val futures = listeners.values.map( listener =>
+        executor.submit((() => acceptConnections(connectionQuotas, listener, numConnections)): Runnable) )
+      futures.foreach(_.get(10, TimeUnit.SECONDS))
+      listeners.values.foreach { listener =>
+        assertEquals(s"${listener.listenerName.value()}",numConnections, connectionQuotas.get(listener.defaultClientIp))
+
+        // verify removing one connection
+        connectionQuotas.dec(listener.listenerName, listener.defaultClientIp)
+        assertEquals(s"Number of connections on $listener:", numConnections - 1, connectionQuotas.get(listener.defaultClientIp))
+
+        // calling dec() for an IP for which we didn't call inc() should throw an exception
+        intercept[IllegalArgumentException](connectionQuotas.dec(listener.listenerName, unknownHost))
```

Comment:
I think that this does not assert as mentioned earlier. Moreover, I think that this deserves its own unit test as it is not really related to the current one.

Reply:
agreed, moved to a separate test